### PR TITLE
Pass Systemd LISTEN_* environment to the container

### PR DESCRIPTION
**Pull request by rhatdan**
If a container is running within a systemd service and it is socket
activated, we need to leak the LISTEN_* environment variables into the
container.

Fixes: https://github.com/containers/podman/issues/10443

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@eriksjolund I would love to work with you on this.  I think this PR is needed to move this forward.

Basically this PR passes the environment variables into the container so PID1 of the container can use them.  We force the LISTEN_PID=1 since systemd will originally set it to the Podman PID.

I am hoping that conmon and crun/runc handle down the open file descriptors correctly.  If you could build a simple test script to test this, I could finish it up.

**Comment by rhatdan** [comment]
@giuseppe PTAL

**Comment by rhatdan** [comment]
@haircommander @giuseppe I believe we have to update the  --preservefds as well when executing conmon.

**Comment by eriksjolund** [comment]
@rhatdan Sure, I'll create a simple test script. I could try it out during the weekend.

File: libpod/oci_conmon_linux.go [hunk 1]
```diff
@@ -1051,6 +1051,16 @@ func (r *ConmonOCIRuntime) createOCIContainer(ctr *Container, restoreOptions *Co
 		}
 	}
 
+	if l, ok := os.LookupEnv("LISTEN_FDS"); ok {
+		listenFds, err := strconv.ParseUint(l, 10, 64)
+		if err != nil {
+			logrus.Warnf("Error LISTEN_FDS environment variable not an int %v", err)
```
Comment (jwhonce) [inline libpod/oci_conmon_linux.go:1057]: Would adding the container id to the warning provide context?

File: cmd/podman/system/service_abi.go [hunk 2]
```diff
@@ -82,6 +82,12 @@ func restService(opts entities.ServiceOptions, flags *pflag.FlagSet, cfg *entiti
 		}
 	}()
 
+	// If the podman service is setup as socket activated, we don't want any containers
+	// to inherit these settings from the service.
+	for _, lEnv := range []string{"LISTEN_PID", "LISTEN_FDS", "LISTEN_FDNAMES"} {
+		os.Unsetenv(lEnv)
```
Comment (jwhonce) [inline cmd/podman/system/service_abi.go:88]: These are also unset in the coreos/go-systemd library when the service is started. See https://github.com/containers/podman/blob/d9eb1269257490dddfae3fde2fe57c3009f94287/vendor/github.com/coreos/go-systemd/v22/activation/files_unix.go#L39 that is called during NewServer...() before Serve()

File: libpod/oci_conmon_linux.go (hunk 1 above)
Comment (rhatdan) [reply libpod/oci_conmon_linux.go:1057]: Added

File: cmd/podman/system/service_abi.go (hunk 2 above)
Comment (rhatdan) [reply cmd/podman/system/service_abi.go:88]: Except I don't see activation.Files(true) ever being called.

Comment (jwhonce) [reply cmd/podman/system/service_abi.go:88]: From `activation.Listeners()` which is called in NewServer...()

https://github.com/containers/podman/blob/d9eb1269257490dddfae3fde2fe57c3009f94287/vendor/github.com/coreos/go-systemd/v22/activation/listeners.go#L29

**Comment by rhatdan** [comment]
@eriksjolund PTAL

**Comment by rhatdan** [comment]
@giuseppe @mheon @vrothberg @edsantiago PTAL, Anyone have a simple test to see if this works?

**Comment by eriksjolund** [comment]
Just some thoughts. (Disclaimer: sending this quite late in the evening so it might not be crystal clear).

_crun_  takes the LISTEN_FDS information from the environment
```
grep -r LISTEN_FDS .
./src/create.c:  if (getenv ("LISTEN_FDS"))
./src/create.c:    crun_context.preserve_fds += strtoll (getenv ("LISTEN_FDS"), NULL, 10);
./src/run.c:  if (getenv ("LISTEN_FDS"))
./src/run.c:    crun_context.preserve_fds += strtoll (getenv ("LISTEN_FDS"), NULL, 10);
./src/exec.c:  if (getenv ("LISTEN_FDS"))
./src/exec.c:    crun_context.preserve_fds += strtoll (getenv ("LISTEN_FDS"), NULL, 10);
```
That looks good.

I think we need to keep the preserve_fds information and LISTEN_FDS information apart in the Podman and Conmon code.

In the source code for Podman and Conmon the term preserve_fds could for instance be "reserved" to always mean the user-provided argument __--preserve-fds__ from `podman run` and `podman exec`. It would be a constant value  that would not change.

One thing to think about

Quote from: https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html
```
#define SD_LISTEN_FDS_START 3
```

In other words. The first LISTEN_FDS file descriptor needs to come directly after stdin, stdout, stderr.

File: libpod/oci_conmon_linux.go [hunk 3]
```diff
@@ -1055,6 +1055,16 @@ func (r *ConmonOCIRuntime) createOCIContainer(ctr *Container, restoreOptions *Co
...
 
+	if l, ok := os.LookupEnv("LISTEN_FDS"); ok {
+		listenFds, err := strconv.ParseUint(l, 10, 64)
+		if err != nil {
+			logrus.Warnf("Error LISTEN_FDS environment variable for %s not an int %v", ctr.ID(), err)
+		}
+
+		if uint(listenFds) > ctr.config.PreserveFDs {
+			ctr.config.PreserveFDs = uint(listenFds)
+		}
```
Comment (eriksjolund) [inline libpod/oci_conmon_linux.go:1066]:
```
		if uint(listenFds) > ctr.config.PreserveFDs {
			ctr.config.PreserveFDs = uint(listenFds)
		}
```
seems to be 
(pseudo code)
```
ctr.config.PreserveFDs := max(listenFds, ctr.config.PreserveFDs)
```

I don't understand this code. Why should it be a max value?

Comment (rhatdan) [reply libpod/oci_conmon_linux.go:1066]: I am just guessing here, that if the user specified --preserve_fds=1 and LISTEN_FDS=2, then we need to leak at least 2 FDs into the container.  But I am not really sure what should be the behaviour.  Perhaps we should ignore the LISTEN_FDS and force the user to tell us how many FDS to leak.

**Comment by rhatdan** [comment]
I don't want to add another option about leaked FDs into conmon, since conmon would not know which one to take priority with.
@giuseppe @vrothberg @haircommander WDYT?

File: libpod/oci_conmon_linux.go (hunk 3 above)
Comment (eriksjolund) [reply libpod/oci_conmon_linux.go:1066]: At first sight runc and crun are doing the same:
See the runc example:
```
 LISTEN_PID=$pid_of_runc LISTEN_FDS=3 runc run --preserve-fds 5 <container>
```

from https://github.com/opencontainers/runc/blob/master/docs/terminals.md

-------------


Quote: _basically, it can just leave the FDs and $LISTEN_FDS untouched_
https://systemd.io/CONTAINER_INTERFACE/

I guess runc and crun could have ignored LISTEN_FDS and just make it the responsibility of 
the caller to specify a --preserve-fds  argument that includes the LISTEN_FDS.

But that is not how it is done right now.

Comment (rhatdan) [reply libpod/oci_conmon_linux.go:1066]: So we should do nothing with them except leak them into the OCI Runtimes.  Will crun and runc rename the LISTEN_PID to pid1?

**Comment by eriksjolund** [comment]
@rhatdan Regarding the test scripts, I gave it a try but I haven't finished it yet.
I wanted to do a simple server that echoed the input back.

Here is some code that might be useful if anyone is trying to do the same
```
$ egrep -r 'echo.sock|5555' . | grep service
./oneshot-test.service:ExecStart=bash -c "echo abc | socat tcp:localhost:5555 - > %t/reply.txt"
./echo.service:#ExecStart=socat TCP4-LISTEN:5555,fork EXEC:/bin/cat
./echo.service:ExecStart=socat UNIX-LISTEN:%t/echo.sock,fork EXEC:/bin/cat
./socket-activated-echo.service: #ExecStart=/usr/lib/systemd/systemd-socket-proxyd localhost:5555
./socket-activated-echo.service: ExecStart=/usr/lib/systemd/systemd-socket-proxyd %t/echo.sock
```
I don't remember what was not working. (Adding `podman` to the commands would be the next step).

**Comment by rhatdan** [comment]
@eriksjolund any progress?

**Comment by eriksjolund** [comment]
@rhatdan No, I have been busy with other things.

File: libpod/oci_conmon_linux.go (hunk 3 above)
Comment (vrothberg) [reply libpod/oci_conmon_linux.go:1066]: That's correct.  crun and runc preserve the FDS but only runc sets LISTEN_PID=1 (Cc: @giuseppe @flouthoc).

**Comment by vrothberg** [comment]
Closing, @rhatdan aksed me to take over.
